keps/sig-node/20191205-container-streaming-requests.md

```diff
@@ -0,0 +1,222 @@
+complicates configuration, and opens the cluster to potential security risks. This is a proposal to
+consolidate the current options into a single supported configuration. Specifically, it proposes:
+
+1. Graduate `ValidateProxyRedirects` to GA
+2. Deprecate the Kubelet's `--redirect-container-streaming` feature
+3. Deprecate the `StreamingProxyRedirects` feature
+
+## Motivation
+
+The `StreamingProxyRedirects` feature was originally added as part of the migration the [Container
+Runtime Interface][]. One of the initial goals was to bypass the Kubelet for these long-running
+requests in order to better account the request overhead. However, the apiserver is still
+responsible for proxying these requests, so the gains of bypassing the Kubelet are
+marginal. Furthermore, the redirect handling significantly complicates the apiserver stream
+handling, and opens it up to [potential SSRF attacks][], so we added the `ValidateProxyRedirects`
+feature to mitigate those.
+
+When we went to add mTLS support to the runtime streaming server, we realized getting the right
+certs established between the apiserver and the runtime was complicated, but the apiserver already
+had an mTLS connection to the Kubelet. So, we added the ability for the Kubelet to connect the
+stream locally and proxy the connection back to the apiserver. This setup is more secure, removes
+the need for `StreamingProxyRedirects` and `ValidateProxyRedirects`, but adds the Kubelet back into
+the long-running request chain.
+
+`StreamingProxyRedirects` and `ValidateProxyRedirects` are currently in Beta, and should either be
+graduated or removed.
+
+[Container Runtime Interface]: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/container-runtime-interface-v1.md
+[potential SSRF attacks]: https://github.com/kubernetes/kubernetes/issues/85867
+
+### Goals
+
+1. Simplify secure container runtime setup and operation
+2. Remove conformance dependencies on beta features
+
+### Non-Goals
+
+Design a high-bandwidth low-latency production ready streaming channel to containers. Such a system
+should go through an ingress point other than the apiserver, and could be implemented (or at least
+prototyped) out of tree by leveraging the new [ephemeral containers
+feature](20190212-ephemeral-containers.md).
+
+## Proposal
+
+After the [transition period](#rollout-plan), the only way to configure container streaming will be
+using the local redirect with Kubelet proxy (`--redirect-container-streaming=false`) approach. Under
+this configuration, handling streaming redirects are not needed by the apiserver, so
+`StreamingProxyRedirects` will be deprecated and eventually removed. Once `StreamingProxyRedirects`
+are removed, `ValidateProxyRedirects` is no longer needed and can be removed. However, it will be
+faster to graduate that feature to GA first, and then it can be removed along with the
```

Comment:
Why we need to graduate it to GA if we plan to remove it? I don't quite get the logic here. :p

Reply:
We don't strictly need to. It's already enabled by default. But, removal of the code along with StreamingProxyRedirects won't happen until 1.22 (or 24). The idea of going to GA is that we can clean it up (and not depend on a beta feature) for the 4-6 releases before it goes away.

Reply:
I wouldn't promote it and deprecate it at the same time… that's pretty weird

Reply:
OK. Promotion in this case really just means cleaning up the feature gate, but I suppose that's not necessary.